File: test_samples/Sample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;

namespace AutoFmea.Services
{
    /// <summary>
    /// UserService — Manages user accounts and authentication.
    /// Depended on by AuthController, SessionManager, and ReportService.
    /// </summary>
    public class UserService
    {
        private readonly IUserRepository _userRepository;
        private readonly ILogger<UserService> _logger;
        private readonly int _sessionTtl;

        public UserService(IUserRepository userRepository, ILogger<UserService> logger)
        {
            _userRepository = userRepository ?? throw new ArgumentNullException(nameof(userRepository));
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
            _sessionTtl = 3600;
        }

        public async Task<User?> GetUserAsync(int userId)
        {
            if (userId <= 0)
                throw new ArgumentException("User ID must be positive.", nameof(userId));

            try
            {
                return await _userRepository.FindByIdAsync(userId);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to retrieve user {UserId}", userId);
                throw;
            }
        }

        public async Task<User> CreateUserAsync(string username, string email, string role = "user")
        {
            if (string.IsNullOrWhiteSpace(username) || username.Length < 3)
                throw new ArgumentException("Username must be at least 3 characters.");

            if (!email.Contains('@'))
                throw new ArgumentException("Invalid email address.");

            var validRoles = new[] { "user", "admin", "moderator" };
            if (!validRoles.Contains(role))
                throw new ArgumentException($"Invalid role: {role}");

            if (await _userRepository.ExistsByEmailAsync(email))
                throw new InvalidOperationException("A user with this email already exists.");

            var user = new User
            {
                Username = username,
                Email = email,
                Role = role,
                CreatedAt = DateTime.UtcNow
            };

            return await _userRepository.SaveAsync(user);
        }

        public async Task<bool> DeleteUserAsync(int userId, int requestingUserId)
        {
            var requestor = await GetUserAsync(requestingUserId);
            if (requestor == null || requestor.Role != "admin")
                throw new UnauthorizedAccessException("Only admins can delete users.");

            var target = await GetUserAsync(userId);
            if (target == null) return false;

            await _userRepository.DeleteByIdAsync(userId);
            return true;
        }

        private bool IsValidPassword(string password)
        {
            if (string.IsNullOrEmpty(password) || password.Length < 8) return false;
            bool hasUpper = password.Any(char.IsUpper);
            bool hasDigit = password.Any(char.IsDigit);
            return hasUpper && hasDigit;
        }
    }

    /// <summary>
    /// DataProcessor — Processes raw data records through a validation pipeline.
    /// </summary>
    public class DataProcessor
    {
        private readonly List<string> _errors = new();
        private int _processedCount;

        public List<string> ProcessBatch(List<string> records)
        {
            var results = new List<string>();

            for (int i = 0; i < records.Count; i++)
            {
                var record = records[i];

                if (string.IsNullOrWhiteSpace(record))
                {
                    _errors.Add($"Record {i}: empty record.");
                    continue;
                }

                try
                {
                    var validated = Validate(record);
                    var transformed = Transform(validated);
                    results.Add(transformed);
                    _processedCount++;
                }
                catch (Exception ex)
                {
                    _errors.Add($"Record {i}: {ex.Message}");
                }
            }

            return results;
        }

        private string Validate(string record)
        {
            if (record.Length > 10000)
                throw new ArgumentException("Record exceeds maximum size.");
            return record.Trim();
        }

        private string Transform(string record)
        {
            return record.ToUpperInvariant();
        }

        public string GetReport()
        {
            return $"Processed: {_processedCount}, Errors: {_errors.Count}";
        }
    }

    /// <summary>
    /// SessionManager — Manages user sessions. Used by UserService.
    /// </summary>
    public class SessionManager
    {
        private readonly Dictionary<string, (object Data, DateTime Created)> _store = new();
        private readonly int _ttl;

        public SessionManager(int ttlSeconds = 3600)
        {
            _ttl = ttlSeconds;
        }

        public void Create(string token, object data)
        {
            _store[token] = (data, DateTime.UtcNow);
        }

        public object? Get(string token)
        {
            if (!_store.TryGetValue(token, out var entry)) return null;
            if ((DateTime.UtcNow - entry.Created).TotalSeconds > _ttl)
            {
                _store.Remove(token);
                return null;
            }
            return entry.Data;
        }

        public void Invalidate(string token)
        {
            _store.Remove(token);
        }
    }
}

File: test_samples/Sample.java
package com.example.app;

import com.example.repository.UserRepository;
import com.example.model.User;
import com.example.auth.AuthController;
import java.util.List;
import java.util.Optional;
import java.time.LocalDateTime;

/**
 * UserService — Manages user accounts, authentication, and session handling.
 * This class is depended on by AuthController, SessionManager, and UserRepository.
 */
public class UserService {

    private final UserRepository userRepository;
    private final AuthController authController;
    private int sessionTtl;

    public UserService(UserRepository userRepository, AuthController authController) {
        this.userRepository = userRepository;
        this.authController = authController;
        this.sessionTtl = 3600;
    }

    public Optional<User> getUser(Long userId) {
        if (userId == null || userId <= 0) {
            throw new IllegalArgumentException("User ID must be a positive number.");
        }
        return userRepository.findById(userId);
    }

    public User createUser(String username, String email, String role) {
        if (username == null || username.length() < 3) {
            throw new IllegalArgumentException("Username must be at least 3 characters.");
        }
        if (email == null || !email.contains("@")) {
            throw new IllegalArgumentException("Invalid email address.");
        }
        if (!role.equals("user") && !role.equals("admin") && !role.equals("moderator")) {
            throw new IllegalArgumentException("Invalid role: " + role);
        }
        if (userRepository.existsByEmail(email)) {
            throw new RuntimeException("A user with this email already exists.");
        }
        User newUser = new User(username, email, role, LocalDateTime.now());
        return userRepository.save(newUser);
    }

    public boolean deleteUser(Long userId, Long requestingUserId) {
        Optional<User> requestor = getUser(requestingUserId);
        if (!requestor.isPresent() || !requestor.get().getRole().equals("admin")) {
            throw new SecurityException("Only admins can delete users.");
        }
        Optional<User> target = getUser(userId);
        if (!target.isPresent()) {
            return false;
        }
        userRepository.deleteById(userId);
        return true;
    }

    public List<User> listUsers(String roleFilter, int pageSize, int page) {
        if (pageSize <= 0 || pageSize > 100) {
            pageSize = 20;
        }
        if (page < 0) {
            page = 0;
        }
        if (roleFilter != null && !roleFilter.isEmpty()) {
            return userRepository.findByRole(roleFilter, page, pageSize);
        }
        return userRepository.findAll(page, pageSize);
    }

    private boolean isValidPassword(String password) {
        if (password == null || password.length() < 8) return false;
        boolean hasUpper = false, hasDigit = false;
        for (char c : password.toCharArray()) {
            if (Character.isUpperCase(c)) hasUpper = true;
            if (Character.isDigit(c)) hasDigit = true;
        }
        return hasUpper && hasDigit;
    }
}

class DataProcessor {
    private List<String> errors;
    private int processedCount;

    public DataProcessor() {
        this.errors = new java.util.ArrayList<>();
        this.processedCount = 0;
    }

    public List<String> processBatch(List<String> records) {
        List<String> results = new java.util.ArrayList<>();
        for (int i = 0; i < records.size(); i++) {
            String record = records.get(i);
            if (record == null || record.isEmpty()) {
                errors.add("Record " + i + ": empty record.");
                continue;
            }
            try {
                String validated = validate(record);
                String transformed = transform(validated);
                results.add(transformed);
                processedCount++;
            } catch (Exception e) {
                errors.add("Record " + i + ": " + e.getMessage());
            }
        }
        return results;
    }

    private String validate(String record) {
        if (record.length() > 10000) {
            throw new IllegalArgumentException("Record exceeds maximum size.");
        }
        return record.trim();
    }

    private String transform(String record) {
        return record.toUpperCase().replaceAll("\\s+", " ");
    }

    public String getReport() {
        return String.format("Processed: %d, Errors: %d", processedCount, errors.size());
    }
}
